dlib/dnn/trainer.h

```diff
@@ -1057,34 +1081,39 @@ namespace dlib
             if (!std::ifstream(newest_syncfile(), std::ios::binary))
                 return false;
 
-            for (auto x : previous_loss_values)
+            // if we haven't seen much data yet then just say false.
+            if (gradient_updates_since_last_sync < 30)
+                return false;
+
+            // Now look at the data since a little before the last disk sync.  We will
+            // check if the loss is getting better or worse.
+            while (previous_loss_values_to_keep_until_disk_sync.size() > 2 * gradient_updates_since_last_sync)
+                previous_loss_values_to_keep_until_disk_sync.pop_front();
+
+            for (auto x : previous_loss_values_to_keep_until_disk_sync)
             {
                 // If we get a NaN value of loss assume things have gone horribly wrong and
                 // we should reload the state of the trainer.
                 if (std::isnan(x))
                     return true;
             }
 
-            // if we haven't seen much data yet then just say false.  Or, alternatively, if
-            // it's been too long since the last sync then don't reload either.
-            if (gradient_updates_since_last_sync < 30 || previous_loss_values.size() < 2*gradient_updates_since_last_sync)
-                return false;
-
-            // Now look at the data since a little before the last disk sync.  We will
-            // check if the loss is getting bettor or worse.
             running_gradient g;
-            for (size_t i = previous_loss_values.size() - 2*gradient_updates_since_last_sync; i < previous_loss_values.size(); ++i)
-                g.add(previous_loss_values[i]);
+            for (size_t i = 0; i < previous_loss_values_to_keep_until_disk_sync.size(); ++i)
+                g.add(previous_loss_values_to_keep_until_disk_sync[i]);
```

Comment:
Just make this part of the above loop that does the NaN check.  No need to repeat the loop over `previous_loss_values_to_keep_until_disk_sync` again.

Reply:
Fixed.